File: sdk/typescript/src/orchestrator.ts
/**
 * Orchestrator client implementation for the Aether TypeScript SDK.
 *
 * Orchestrators manage agent/task lifecycle: receiving startup requests
 * when targeted agents are offline, launching compute resources, and
 * managing agent pools and scaling.
 *
 * @module orchestrator
 */

import { AetherClient } from "./client.js";
import type { AetherClientOptions } from "./client.js";
import { MessageType } from "./types.js";
import type { TaskAssignment, ConnectionAck } from "./types.js";
import { InvalidArgumentError } from "./errors.js";
import {
  agentTopic,
  uniqueTaskTopic,
  taskBroadcastTopic,
} from "./topics.js";

// =============================================================================
// Orchestrator Client Options
// =============================================================================

/**
 * Configuration options for the OrchestratorClient.
 */
export interface OrchestratorClientOptions extends AetherClientOptions {
  /** The orchestrator implementation type (e.g., "kubernetes-orchestrator"). Required. */
  implementation: string;
  /** Profiles this orchestrator supports (e.g., ["kubernetes", "docker"]). Required, at least one. */
  supportedProfiles: string[];
  /** Unique specifier for this instance. Auto-generated if not provided. */
  specifier?: string;
}

// =============================================================================
// OrchestratorClient
// =============================================================================

/**
 * Client for connecting to the Aether gateway as an orchestrator.
 *
 * Orchestrators receive task assignments via the onTaskAssignment handler
 * and are responsible for launching the appropriate compute resources.
 *
 * @example
 * ```typescript
 * const orchestrator = new OrchestratorClient({
 *   address: "localhost:50051",
 *   implementation: "k8s-orchestrator",
 *   supportedProfiles: ["kubernetes"],
 * });
 *
 * orchestrator.onTaskAssignment((assignment) => {
 *   console.log(`Starting ${assignment.targetImplementation} for task ${assignment.taskId}`);
 *   // Launch container based on assignment.launchParams
 * });
 *
 * await orchestrator.connect();
 * ```
 */
export class OrchestratorClient extends AetherClient {
  private readonly _implementation: string;
  private readonly _specifier: string;
  private readonly _supportedProfiles: string[];

  constructor(options: OrchestratorClientOptions) {
    super(options);

    if (!options.implementation) {
      throw new InvalidArgumentError("Implementation is required", "implementation");
    }
    if (!options.supportedProfiles || options.supportedProfiles.length === 0) {
      throw new InvalidArgumentError("At least one supported profile is required", "supportedProfiles");
    }

    this._implementation = options.implementation;
    this._specifier = options.specifier ?? crypto.randomUUID().slice(0, 8);
    this._supportedProfiles = options.supportedProfiles;
  }

  // ===========================================================================
  // Identity Accessors
  // ===========================================================================

  get implementation(): string {
    return this._implementation;
  }

  get specifier(): string {
    return this._specifier;
  }

  get supportedProfiles(): string[] {
    return [...this._supportedProfiles];
  }

  // ===========================================================================
  // Init Message
  // ===========================================================================

  /** @internal */
  protected override _buildInitMessage(): Record<string, unknown> {
    return {
      orchestrator: {
        implementation: this._implementation,
        specifier: this._specifier,
        supportedProfiles: this._supportedProfiles,
      },
      credentials: this._credentials,
      resumeSessionId: this._resumeSessionId,
    };
  }

  // ===========================================================================
  // Message Sending Helpers
  // ===========================================================================

  /**
   * Sends a status/control message to an agent.
   */
  sendStatusToAgent(
    workspace: string,
    implementation: string,
    specifier: string,
    payload: Uint8Array,
    messageType: MessageType = MessageType.Control,
  ): void {
    this._sendMessage(agentTopic(workspace, implementation, specifier), payload, messageType);
  }

  /**
   * Sends a status/control message to a task.
   * For unique tasks (with specifier), uses tu.* topic.
   * For non-unique tasks (empty specifier), uses tb.* broadcast topic.
   */
  sendStatusToTask(
    workspace: string,
    implementation: string,
    specifier: string,
    payload: Uint8Array,
    messageType: MessageType = MessageType.Control,
  ): void {
    const topic = specifier
      ? uniqueTaskTopic(workspace, implementation, specifier)
      : taskBroadcastTopic(workspace, implementation);
    this._sendMessage(topic, payload, messageType);
  }
}

// =============================================================================
// BaseOrchestrator Options
// =============================================================================

/**
 * Configuration options for BaseOrchestrator subclasses.
 */
export interface BaseOrchestratorOptions extends OrchestratorClientOptions {
  /**
   * Whether to log task assignment details to the console.
   * Default: false.
   */
  logAssignments?: boolean;
}

// =============================================================================
// BaseOrchestrator
// =============================================================================

/**
 * Abstract base class for building orchestrators.
 *
 * BaseOrchestrator wraps {@link OrchestratorClient} and provides a structured
 * framework for managing agent/task lifecycle. Subclasses implement the
 * abstract {@link launchTask} method to handle task assignments with their
 * specific compute backend (Docker, Kubernetes, subprocess, etc.).
 *
 * The class manages the full orchestrator lifecycle:
 * - Connecting to the Aether gateway as an Orchestrator principal
 * - Receiving {@link TaskAssignment} messages and dispatching to launchTask
 * - Error handling and logging hooks
 *
 * @example
 * ```typescript
 * import { BaseOrchestrator, TaskAssignment } from "@scitrera/aether-client";
 *
 * class MyOrchestrator extends BaseOrchestrator {
 *   async launchTask(assignment: TaskAssignment): Promise<void> {
 *     console.log(`Launching task ${assignment.taskId} (profile: ${assignment.profile})`);
 *     // Start a subprocess, container, etc. using assignment.launchParams
 *   }
 * }
 *
 * const orch = new MyOrchestrator({
 *   address: "localhost:50051",
 *   implementation: "my-orchestrator",
 *   supportedProfiles: ["my-profile"],
 * });
 *
 * orch.onConnect((ack) => {
 *   console.log(`Connected with session ${ack.sessionId}`);
 * });
 *
 * await orch.connect();
 * ```
 */
export abstract class BaseOrchestrator {
  /** The underlying OrchestratorClient for direct gateway access. */
  readonly client: OrchestratorClient;

  private readonly _logAssignments: boolean;

  /**
   * Creates a new BaseOrchestrator.
   *
   * @param options - Orchestrator configuration options
   */
  constructor(options: BaseOrchestratorOptions) {
    this._logAssignments = options.logAssignments ?? false;
    this.client = new OrchestratorClient(options);

    // Wire task assignment handler to launchTask
    this.client.onTaskAssignment((assignment) => {
      return this._handleAssignment(assignment);
    });
  }

  // ===========================================================================
  // Abstract Interface
  // ===========================================================================

  /**
   * Called when the gateway assigns a task to this orchestrator.
   *
   * Subclasses must implement this method to launch the appropriate compute
   * resource (container, process, VM, etc.) based on the assignment details.
   *
   * The assignment includes:
   * - `taskId`: Unique task identifier
   * - `profile`: The profile that matched (from supportedProfiles)
   * - `targetImplementation`: The agent/task implementation to launch
   * - `workspace`: The workspace this task belongs to
   * - `specifier`: Optional unique specifier for the agent/task
   * - `launchParams`: Key-value launch parameters (image, command, env vars, etc.)
   * - `metadata`: Arbitrary task metadata
   *
   * @param assignment - The task assignment received from the gateway
   */
  abstract launchTask(assignment: TaskAssignment): void | Promise<void>;

  // ===========================================================================
  // Lifecycle
  // ===========================================================================

  /**
   * Connects to the Aether gateway.
   *
   * Must be called before the orchestrator can receive task assignments.
   */
  async connect(): Promise<void> {
    await this.client.connect();
  }

  /**
   * Disconnects from the Aether gateway.
   */
  async disconnect(): Promise<void> {
    await this.client.disconnect();
  }

  /**
   * Returns whether the orchestrator is currently connected.
   */
  get connected(): boolean {
    return this.client.connected;
  }

  /**
   * Returns the current session ID assigned by the gateway.
   * Empty string if not connected.
   */
  get sessionId(): string {
    return this.client.sessionId;
  }

  /**
   * Returns the orchestrator's implementation type.
   */
  get implementation(): string {
    return this.client.implementation;
  }

  /**
   * Returns the orchestrator's specifier.
   */
  get specifier(): string {
    return this.client.specifier;
  }

  /**
   * Returns the list of profiles this orchestrator supports.
   */
  get supportedProfiles(): string[] {
    return this.client.supportedProfiles;
  }

  // ===========================================================================
  // Handler Registration
  // ===========================================================================

  /**
   * Registers a handler for successful connection events.
   *
   * @param handler - Called with the ConnectionAck when connected
   */
  onConnect(handler: (ack: ConnectionAck) => void | Promise<void>): void {
    this.client.onConnect(handler);
  }

  /**
   * Registers a handler for disconnection events.
   *
   * @param handler - Called with the reason string when disconnected
   */
  onDisconnect(handler: (reason: string) => void | Promise<void>): void {
    this.client.onDisconnect(handler);
  }

  /**
   * Registers a handler for reconnection attempts.
   *
   * @param handler - Called with the attempt number on each reconnect try
   */
  onReconnecting(handler: (attempt: number) => void | Promise<void>): void {
    this.client.onReconnecting(handler);
  }

  // ===========================================================================
  // Message Sending
  // ===========================================================================

  /**
   * Sends a status/control message to an agent.
   *
   * @param workspace - Target agent's workspace
   * @param implementation - Target agent's implementation type
   * @param specifier - Target agent's specifier
   * @param payload - Message payload
   * @param messageType - Message type. Default: Control
   */
  sendStatusToAgent(
    workspace: string,
    implementation: string,
    specifier: string,
    payload: Uint8Array,
    messageType: MessageType = MessageType.Control,
  ): void {
    this.client.sendStatusToAgent(workspace, implementation, specifier, payload, messageType);
  }

  /**
   * Sends a status/control message to a task.
   *
   * @param workspace - Target task's workspace
   * @param implementation - Target task's implementation type
   * @param specifier - Target task's specifier (empty for broadcast)
   * @param payload - Message payload
   * @param messageType - Message type. Default: Control
   */
  sendStatusToTask(
    workspace: string,
    implementation: string,
    specifier: string,
    payload: Uint8Array,
    messageType: MessageType = MessageType.Control,
  ): void {
    this.client.sendStatusToTask(workspace, implementation, specifier, payload, messageType);
  }

  // ===========================================================================
  // Internal
  // ===========================================================================

  /**
   * Internal handler that logs and dispatches to launchTask.
   * @internal
   */
  private async _handleAssignment(assignment: TaskAssignment): Promise<void> {
    if (this._logAssignments) {
      console.log(
        `[BaseOrchestrator] Task assignment received: taskId=${assignment.taskId}` +
          ` profile=${assignment.profile} impl=${assignment.targetImplementation}` +
          ` workspace=${assignment.workspace}`,
      );
    }
    await this.launchTask(assignment);
  }
}

File: sdk/go/aether/orchestrator.go
// Package aether orchestrator client implementation.
//
// This file provides the OrchestratorClient type for connecting to the Aether
// gateway as an orchestrator. Orchestrators are responsible for managing
// agent/task lifecycle, including launching compute resources and handling
// task assignments.

package aether

import (
	pb "github.com/scitrera/aether/api/proto"
)

// =============================================================================
// Orchestrator Client
// =============================================================================

// OrchestratorClient is a client for connecting to the Aether gateway as an orchestrator.
//
// Orchestrators are responsible for managing agent/task lifecycle:
//   - Receiving startup requests when targeted agents are offline
//   - Launching compute resources (containers, VMs, etc.)
//   - Managing agent pools and scaling
//   - Processing task assignments based on supported profiles
//
// Orchestrators register with a list of profiles they support. When the gateway
// needs to start an agent matching one of these profiles, it sends a TaskAssignment
// to the appropriate orchestrator.
//
// OrchestratorClient embeds BaseClient and adds orchestrator-specific functionality:
//   - Profile management (list of supported profiles)
//   - Task assignment handling via OnTaskAssignment callback
//   - Status/control message sending to agents and tasks
//
// Example usage:
//
//	client, err := aether.NewOrchestratorClient(aether.OrchestratorOptions{
//	    ClientOptions: aether.ClientOptions{
//	        ServerAddr: "localhost:50051",
//	    },
//	    Implementation:    "kubernetes-orchestrator",
//	    SupportedProfiles: []string{"k8s-worker", "k8s-agent"},
//	})
//	if err != nil {
//	    log.Fatal(err)
//	}
//
//	client.OnTaskAssignment(func(ctx context.Context, assignment *aether.TaskAssignment) error {
//	    fmt.Printf("Received task assignment: %s (profile: %s)\n",
//	        assignment.TaskID, assignment.Profile)
//	    // Launch the appropriate compute resource based on the assignment
//	    return nil
//	})
//
//	if err := client.Connect(ctx); err != nil {
//	    log.Fatal(err)
//	}
//
//	// Start the message loop (blocks until disconnection)
//	if err := client.Run(ctx); err != nil {
//	    log.Fatal(err)
//	}
type OrchestratorClient struct {
	*BaseClient

	// Identity fields
	implementation    string
	specifier         string   // Optional; server generates if empty
	supportedProfiles []string // Required; at least one profile
	credentials       map[string]string
}

// NewOrchestratorClient creates a new OrchestratorClient with the given options.
//
// The client is created but not connected. Call Connect() to establish
// the connection to the server.
//
// Returns an error if required options are missing (ServerAddr, Implementation,
// at least one SupportedProfile).
func NewOrchestratorClient(opts OrchestratorOptions) (*OrchestratorClient, error) {
	// Validate options
	if err := opts.Validate(); err != nil {
		return nil, err
	}

	// Create base client configuration
	cfg := BaseClientConfig{
		ServerAddr:  opts.ServerAddr,
		Connection:  opts.Connection,
		TLS:         opts.TLS,
		Credentials: opts.Credentials,
	}

	// Create base client
	base, err := NewBaseClient(cfg)
	if err != nil {
		return nil, err
	}

	// Copy supported profiles slice to avoid aliasing issues
	profiles := make([]string, len(opts.SupportedProfiles))
	copy(profiles, opts.SupportedProfiles)

	// Create orchestrator client
	oc := &OrchestratorClient{
		BaseClient:        base,
		implementation:    opts.Implementation,
		specifier:         opts.Specifier,
		supportedProfiles: profiles,
		credentials:       opts.Credentials,
	}

	// Set the init message builder
	base.initMsgBuilder = oc.buildInitMessage

	return oc, nil
}

// buildInitMessage creates the InitConnection message for orchestrator identity.
func (c *OrchestratorClient) buildInitMessage() *pb.InitConnection {
	return &pb.InitConnection{
		ClientType: &pb.InitConnection_Orchestrator{
			Orchestrator: &pb.OrchestratorIdentity{
				Implementation:    c.implementation,
				Specifier:         c.specifier,
				SupportedProfiles: c.supportedProfiles,
			},
		},
		Credentials: c.credentials,
	}
}

// =============================================================================
// Identity Accessors
// =============================================================================

// Implementation returns the orchestrator's implementation type.
//
// Example: "kubernetes-orchestrator", "docker-orchestrator"
func (c *OrchestratorClient) Implementation() string {
	return c.implementation
}

// Specifier returns the orchestrator's specifier (unique instance identifier).
//
// If no specifier was provided during creation, this may be empty until
// the server assigns one upon connection.
func (c *OrchestratorClient) Specifier() string {
	return c.specifier
}

// SupportedProfiles returns a copy of the profiles this orchestrator supports.
//
// Profiles define which types of agents/tasks this orchestrator can spawn.
// The gateway routes task assignments to orchestrators based on matching profiles.
func (c *OrchestratorClient) SupportedProfiles() []string {
	// Return a copy to prevent external modification
	profiles := make([]string, len(c.supportedProfiles))
	copy(profiles, c.supportedProfiles)
	return profiles
}

// SupportsProfile returns true if this orchestrator supports the given profile.
func (c *OrchestratorClient) SupportsProfile(profile string) bool {
	for _, p := range c.supportedProfiles {
		if p == profile {
			return true
		}
	}
	return false
}

// =============================================================================
// Message Sending Helpers
// =============================================================================

// SendStatusToAgent sends a status/control message to a specific agent.
//
// This is typically used to communicate orchestration status updates,
// such as startup confirmations, health checks, or shutdown notices.
//
// Parameters:
//   - workspace: Target agent's workspace
//   - implementation: Target agent's implementation type
//   - specifier: Target agent's specifier
//   - payload: Message payload (bytes)
//
// Uses CONTROL message type. Use SendStatusToAgentWithType for other types.
func (c *OrchestratorClient) SendStatusToAgent(workspace, implementation, specifier string, payload []byte) error {
	return c.SendStatusToAgentWithType(workspace, implementation, specifier, payload, pb.MessageType_CONTROL)
}

// SendStatusToAgentWithType sends a status message to a specific agent with a custom message type.
func (c *OrchestratorClient) SendStatusToAgentWithType(workspace, implementation, specifier string, payload []byte, msgType pb.MessageType) error {
	topic := AgentTopic(workspace, implementation, specifier)
	return c.sendMessage(topic, payload, msgType)
}

// SendStatusToTask sends a status/control message to a specific task.
//
// For unique tasks (with specifier):
//   - Uses tu::{workspace}::{implementation}::{specifier} topic
//
// For non-unique tasks (empty specifier):
//   - Uses tb::{workspace}::{implementation} broadcast topic
//
// Parameters:
//   - workspace: Target task's workspace
//   - implementation: Target task's implementation type
//   - specifier: Target task's specifier (empty for broadcast to non-unique tasks)
//   - payload: Message payload (bytes)
//
// Uses CONTROL message type. Use SendStatusToTaskWithType for other types.
func (c *OrchestratorClient) SendStatusToTask(workspace, implementation, specifier string, payload []byte) error {
	return c.SendStatusToTaskWithType(workspace, implementation, specifier, payload, pb.MessageType_CONTROL)
}

// SendStatusToTaskWithType sends a status message to a specific task with a custom message type.
func (c *OrchestratorClient) SendStatusToTaskWithType(workspace, implementation, specifier string, payload []byte, msgType pb.MessageType) error {
	var topic string
	if specifier != "" {
		// Unique task
		topic = UniqueTaskTopic(workspace, implementation, specifier)
	} else {
		// Non-unique task broadcast
		topic = TaskBroadcastTopic(workspace, implementation)
	}
	return c.sendMessage(topic, payload, msgType)
}

// =============================================================================
// Generic Message Sending
// =============================================================================

// SendMessage sends a message to a specified topic.
//
// This is the low-level sending method. For most use cases, prefer the
// type-specific helpers (SendStatusToAgent, SendStatusToTask, etc.).
//
// Note: Per the Aether specification, orchestrators primarily send status
// updates to agents and tasks. They typically do not send user-facing
// messages or publish events/metrics.
//
// Parameters:
//   - targetTopic: The destination topic string
//   - payload: Message payload (bytes)
//   - msgType: Message type (CHAT, CONTROL, TOOL_CALL, EVENT, METRIC)
func (c *OrchestratorClient) SendMessage(targetTopic string, payload []byte, msgType pb.MessageType) error {
	return c.sendMessage(targetTopic, payload, msgType)
}

// SendControlMessage sends a CONTROL message to a specified topic.
func (c *OrchestratorClient) SendControlMessage(targetTopic string, payload []byte) error {
	return c.sendMessage(targetTopic, payload, pb.MessageType_CONTROL)
}

// SendChatMessage sends a CHAT message to a specified topic.
func (c *OrchestratorClient) SendChatMessage(targetTopic string, payload []byte) error {
	return c.sendMessage(targetTopic, payload, pb.MessageType_OPAQUE)
}
